.github/workflows/test.yml

```diff
@@ -50,3 +50,117 @@ jobs:
+    env:
+      # Export environment variables for all stages.
+      DOCKER_CLI_EXPERIMENTAL: enabled  # for 'docker buildx'
+      DOCKER_BASE: mvdan/shfmt
+      DOCKER_PASSWORD: ${{secrets.DOCKER_PASSWORD}}
+      # We use all platforms for which FROM images in our Dockerfile are
+      # available.
+      DOCKER_PLATFORMS: >
+        linux/386
+        linux/amd64
+        linux/arm/v6
+        linux/arm/v7
+        linux/arm64/v8
+        linux/ppc64le
+        linux/s390x
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v1
+      with:
+        fetch-depth: 1  # no history, just HEAD
+    - name: Set up image tags
+      run: |
+        set -vx
+        # Export environment variable for later stages.
+        if echo "$GITHUB_REF" | grep -q '^refs/heads/'; then
+          # Pushes to (master) branch - deploy 'latest'.
+          echo "::set-env name=TAGS::latest"
+        elif echo "$GITHUB_REF" | grep -q '^refs/tags/'; then
+          # Pushes tag - deploy tag name.
+          echo "::set-env name=TAGS::${GITHUB_REF/refs\/tags\//}"
+        fi
+    - name: Install Docker buildx
+      run: |
+        set -vx
+        # Install up-to-date version of docker, with buildx support.
+        docker_apt_repo='https://download.docker.com/linux/ubuntu'
+        curl -fsSL "${docker_apt_repo}/gpg" | sudo apt-key add -
+        os="$(lsb_release -cs)"
+        sudo add-apt-repository "deb [arch=amd64] $docker_apt_repo $os stable"
+        sudo apt-get update
+        sudo apt-get -y -o Dpkg::Options::="--force-confnew" install docker-ce
+
+        # Enable docker daemon experimental support (for 'pull --platform').
+        config='/etc/docker/daemon.json'
+        if [[ -e "$config" ]]; then
+          sudo sed -i -e 's/{/{ "experimental": true, /' "$config"
+        else
+          echo '{ "experimental": true }' | sudo tee "$config"
+        fi
+        sudo systemctl restart docker
+
+        # Install QEMU multi-architecture support for docker buildx.
+        docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+
+        # Instantiate docker buildx builder with multi-architecture support.
+        docker buildx create --name mybuilder
+        docker buildx use mybuilder
+        # Start up buildx and verify that all is OK.
+        docker buildx inspect --bootstrap
+    - name: Build multi-architecture Docker images with buildx
+      run: |
+        set -vx
+        username=${DOCKER_BASE/\/*/}
+        echo "$DOCKER_PASSWORD" | docker login -u="$username" --password-stdin
+
+        function buildx() {
+          docker buildx build \
+            --platform ${DOCKER_PLATFORMS// /,} \
+            --push \
+            -f cmd/shfmt/Dockerfile \
+            "$@" \
+            .
+        }
+
+        for tag in $TAGS; do
+          buildx -t "$DOCKER_BASE:$tag"
+          buildx -t "$DOCKER_BASE-alpine:$tag" --target alpine
+        done
+    - name: Test multi-architecture Docker images
```

Comment:
I'm not sure what the point of the "test" step is, since it happens after the image is already pushed. If CI fails because of these tests, I assume we shouldn't be pushing at all.

I do think the test is useful for the simplest sanity checks, but it seems more like an informative test at this stage. If that's what you were going for for now, that's fine - but perhaps we can add a TODO about making `docker push` conditional on this step.

Reply:
You are right that the test is past docker push and as such doesn't prevent bad images to get to Docker Hub. It does test if the just pushed images can be pulled and used correctly though, so that's something.

Unfortunately there is no way that I'm aware of at this time to keep multi-architecture docker images in the local docker image cache. You *have to* push the multi-architecture `buildx build` image to Docker Hub in the same command (the `--push` flag). Whenever you have a local version of such an image, you always only have a single-architecture version corresponding to your host architecture. I.e. you also can't pull + retag + push a multi-architecture image without having narrowed down the pushed image to a single architecture.